### PR TITLE
coordinator: Use ARP to detect the gateway rather than ICMP

### PR DESCRIPTION
**Pull request by cyclinder**
## Thanks for contributing!

1 use arp to accelerate rather than icmp
2 some gateway forbid icmp

#### What type of PR is this?

- release/feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes https://github.com/spidernet-io/spiderpool/issues/3535

**Special notes for your reviewer**:

```
{"level":"DEBUG","ts":"2024-06-17T15:53:48.694+0800","logger":"coordinator.new1","caller":"gwconnection/connection.go:78","msg":"[RetryNum: 1]try to send arp request","Action":"ADD","ContainerID":"860b7454527609f2e67224571977d4a18f5c286fb365709fc353abd8ef9bec77","Netns":"/proc/2572117/ns/net","IfName":"eth0","PodName":"macvlan-160-c-848cf7bf6f-lwbcj","PodNamespace":"default"}
{"level":"DEBUG","ts":"2024-06-17T15:53:53.703+0800","logger":"coordinator.new1","caller":"gwconnection/connection.go:217","msg":"Detect gateway: found the response","Action":"ADD","ContainerID":"860b7454527609f2e67224571977d4a18f5c286fb365709fc353abd8ef9bec77","Netns":"/proc/2572117/ns/net","IfName":"eth0","PodName":"macvlan-160-c-848cf7bf6f-lwbcj","PodNamespace":"default","TargetAddress":"fd00:10:7::103"}
{"level":"INFO","ts":"2024-06-17T15:53:53.704+0800","logger":"coordinator.new1","caller":"gwconnection/connection.go:182","msg":"gateway fd00:10:7::103 is reachable, it's located at 00:50:56:b4:27:ca","Action":"ADD","ContainerID":"860b7454527609f2e67224571977d4a18f5c286fb365709fc353abd8ef9bec77","Netns":"/proc/2572117/ns/net","IfName":"eth0","PodName":"macvlan-160-c-848cf7bf6f-lwbcj","PodNamespace":"default"}
{"level":"ERROR","ts":"2024-06-17T15:53:53.705+0800","logger":"coordinator.new1","caller":"gwconnection/connection.go:100","msg":"[RetryNum: 1]failed to resolve: read packet 0a:3a:c7:44:2d:95: i/o timeout","Action":"ADD","ContainerID":"860b7454527609f2e67224571977d4a18f5c286fb365709fc353abd8ef9bec77","Netns":"/proc/2572117/ns/net","IfName":"eth0","PodName":"macvlan-160-c-848cf7bf6f-lwbcj","PodNamespace":"default"}
{"level":"DEBUG","ts":"2024-06-17T15:53:58.706+0800","logger":"coordinator.new1","caller":"gwconnection/connection.go:78","msg":"[RetryNum: 2]try to send arp request","Action":"ADD","ContainerID":"860b7454527609f2e67224571977d4a18f5c286fb365709fc353abd8ef9bec77","Netns":"/proc/2572117/ns/net","IfName":"eth0","PodName":"macvlan-160-c-848cf7bf6f-lwbcj","PodNamespace":"default"}
{"level":"ERROR","ts":"2024-06-17T15:54:03.711+0800","logger":"coordinator.new1","caller":"gwconnection/connection.go:100","msg":"[RetryNum: 2]failed to resolve: read packet 0a:3a:c7:44:2d:95: i/o timeout","Action":"ADD","ContainerID":"860b7454527609f2e67224571977d4a18f5c286fb365709fc353abd8ef9bec77","Netns":"/proc/2572117/ns/net","IfName":"eth0","PodName":"macvlan-160-c-848cf7bf6f-lwbcj","PodNamespace":"default"}

{"level":"DEBUG","ts":"2024-06-17T15:54:08.713+0800","logger":"coordinator.new1","caller":"gwconnection/connection.go:78","msg":"[RetryNum: 3]try to send arp request","Action":"ADD","ContainerID":"860b7454527609f2e67224571977d4a18f5c286fb365709fc353abd8ef9bec77","Netns":"/proc/2572117/ns/net","IfName":"eth0","PodName":"macvlan-160-c-848cf7bf6f-lwbcj","PodNamespace":"default"}
{"level":"INFO","ts":"2024-06-17T15:54:08.714+0800","logger":"coordinator.new1","caller":"gwconnection/connection.go:106","msg":"Gateway 10.7.0.2 is reachable, gateway is located at 00:50:56:b4:27:ca","Action":"ADD","ContainerID":"860b7454527609f2e67224571977d4a18f5c286fb365709fc353abd8ef9bec77","Netns":"/proc/2572117/ns/net","IfName":"eth0","PodName":"macvlan-160-c-848cf7bf6f-lwbcj","PodNamespace":"default"}
```

### CODE REVIEW

File: pkg/networking/gwconnection/connection.go [hunk 1]
```diff
@@ -4,49 +4,216 @@
...
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	gwNetIP := netip.MustParseAddr(string(dg.V4Gw.String()))
+	// we send a gratuitous arp to checking if ip is conflict
+	// we use dad mode(duplicate address detection mode), so
+	// we set source ip to 0.0.0.0
+	packet, err := arp.NewPacket(arp.OperationRequest, ifi.HardwareAddr, netip.MustParseAddr("0.0.0.0"), ethernet.Broadcast, gwNetIP)
```
Comment (weizhoublue) [inline pkg/networking/gwconnection/connection.go:79]: arp probe is forbidden by some brands of gatway
it should use pod ip for source IP

Comment (cyclinder) [reply pkg/networking/gwconnection/connection.go:79]: done

File: pkg/networking/gwconnection/connection.go [hunk 2]
```diff
@@ -5,48 +5,209 @@ package gwconnection
...
+
+	gwNetIP := netip.MustParseAddr(dg.V4Gw.String())
+	packet, err := arp.NewPacket(arp.OperationRequest, ifi.HardwareAddr, netip.MustParseAddr(dg.v4Addr.String()), ethernet.Broadcast, gwNetIP)
+	if err != nil {
+		return err
+	}
+
+	var gwHwAddr string
+	for i := 0; i < dg.retries; i++ {
+		dg.logger.Sugar().Infof("[RetryNum: %v]try to send arp request", i+1)
```
Comment (weizhoublue) [inline pkg/networking/gwconnection/connection.go:83]: Infof  to debug level

File: pkg/networking/gwconnection/connection.go [hunk 3]
```diff
@@ -5,48 +5,209 @@ package gwconnection
...
+		return err
+	}
+
+	client, err := arp.Dial(ifi)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	gwNetIP := netip.MustParseAddr(dg.V4Gw.String())
```
Comment (weizhoublue) [inline pkg/networking/gwconnection/connection.go:75]: is it tested under ipv6 only environment ?

File: cmd/coordinator/cmd/command_add.go [hunk 4]
```diff
@@ -186,21 +187,28 @@ func CmdAdd(args *skel.CmdArgs) (err error) {
...
+				return fmt.Errorf("failed to init the gateway client: %v", err)
+			}
+			p.ParseAddrFromPreresult(prevResult.IPs)
 			for _, gw := range gws {
-				p, err := gwconnection.NewPinger(conf.DetectOptions.Retry, conf.DetectOptions.Interval, conf.DetectOptions.TimeOut, gw, logger)
-				if err != nil {
-					return fmt.Errorf("failed to run NewPinger: %v", err)
+				if gw.To4() != nil {
+					p.V4Gw = gw
+					errg.Go(c.hostNs, c.netns, p.ArpingOverIface)
```
Comment (weizhoublue) [inline cmd/coordinator/cmd/command_add.go:207]: 后边对于 错误码的判断， 只有 errors.Is(err, constant.ErrIPConflict) ，可以添加对于 哪个网关不可导的支持

Comment (cyclinder) [reply cmd/coordinator/cmd/command_add.go:207]: 意义是啥？ 由于是协程并发检测，这里的 err 不仅仅包括 网关检测可能还有 IP 冲突的 error, 如果只是返回错误的提示，目前已经足够了

Comment (cyclinder) [reply cmd/coordinator/cmd/command_add.go:207]: 我明白了，后面还要删IP对吧

File: pkg/networking/gwconnection/connection.go (hunk 2 above)
Comment (cyclinder) [reply pkg/networking/gwconnection/connection.go]: done

**Comment by cyclinder** [comment]
Waiting for https://github.com/spidernet-io/spiderpool/pull/3603 and https://github.com/spidernet-io/spiderpool/pull/3608 merge.

File: cmd/coordinator/cmd/command_add.go (hunk 4 above)
Comment (weizhoublue) [reply cmd/coordinator/cmd/command_add.go:207]: 这个报错能够体现在 describe pod 中，用户能够清晰了解报错的具体内容

Comment (weizhoublue) [reply cmd/coordinator/cmd/command_add.go:207]: does it make sense

Comment (cyclinder) [reply cmd/coordinator/cmd/command_add.go:207]: 如果是网关不可达，报错类似： `failed to resolve: read packet 0a:3a:c7:44:2d:95: i/o timeout`, 这类报错感觉对用户调试没什么帮助

Comment (weizhoublue) [reply cmd/coordinator/cmd/command_add.go:207]: 毕竟在多网卡等场景下，网关很多，不直接报错具体的网关，会增加排障负担

Comment (cyclinder) [reply cmd/coordinator/cmd/command_add.go:207]: 目前的报错会显示报错的网关是什么，反而返回 `failed to resolve: read packet 0a:3a:c7:44:2d:95: i/o timeout` 没有可用信息

Comment (weizhoublue) [reply cmd/coordinator/cmd/command_add.go:207]: 那不如直接 新构建返回给 runtime 具体错误信息

**Review by weizhoublue — APPROVED** [review]

**Pull request merged**